projects/beta_49_ssl_handshake_version_spike.kp/knowledge.md

```diff
@@ -0,0 +1,246 @@
+
+
+
+## Found the Offending Client
+
+Next is to determine the distribution of pings as well as the distribution of measurements. Were these all unique pings, or is this "fake" data similar to the [Aurora 51 client?](http://reports.telemetry.mozilla.org/post/projects%2Fproblematic_client.kp)
+
+
+```python
+cid = max(res, key = lambda x: x[1]['count'])[0]
+
+cid_pings = pings.filter(
+    lambda x: x.get('clientId') == cid
+)
+
+counts = cid_pings.map(
+    lambda x: sum(get_shv(x).get('values', {}).values())
+).collect()
+
+"Client has {} pings".format(cid_pings.count())
+```
+
+
+
+
+    'Client has 232287 pings'
+
+
+
+## Huge Counts per Ping
+
+The below plot shows a histogram of SSL_HANDSHAKE_VERSION measurement counts per ping, where the height of the bar is number of pigns with that number of measurements. The high counts are why we are seeing such a spike, and the clustering indicates that these might be duplicates.
+
+
+```python
+from matplotlib import pyplot as plt
+import seaborn as sns
+
+%matplotlib inline
+
+plt.title("Number of SSL_HANDSHAKE_VERSION Measurements per Ping")
+plt.xlabel("SSL_HANDSHAKE_VERSION Measurements")
+plt.ylabel("Ping Count")
+
+plt.hist(counts, bins=50)
+plt.legend()
+```
+
+
+![png](images/output_9_1.png)
+
+
+## We are looking at duplicate pings
+
+The plot below shows number of pings and number of distinct DocIDs for this single client per day (looking around the problem time period). We are seeing about 10 repeated docIds, all with high counts of SSL_HANDSHAKE_VERSION.
+
+We have started deduping server-side so some of these pings would be removed; deduping will improve over the next few months so that we recieve less than 10% of dupes.
+
+
+```python
+# What does this client's distribution look like across a the bad days?
+# We know bad days from the aggregates data, see
+# https://github.com/fbertsch/telemetry-repro/blob/2bf7e062719ae498c12f15456e94af9ee7758093/output.txt
+
+begin = '20170305'
+end = '20170320'
+
+cid_pings = Dataset \
+    .from_source('telemetry') \
+    .where (
+        appUpdateChannel = 'beta',
+        appVersion = lambda x: x.startswith('49'),
+        submissionDate = lambda x: begin <= x <= end
+    ).records(sc) \
+    .filter(lambda x: x.get('clientId') == cid)
+
+props = {
+    'submission_date': 'meta/submissionDate',
+    'version': 'version',
+    'build_id': 'buildId',
+    'document_id': 'meta/documentId',
+    'timestamp': 'meta/Timestamp',
+    'SSL_HANDSHAKE_VERSION': 'payload/histograms/SSL_HANDSHAKE_VERSION'
+}
+    
+data = get_pings_properties(cid_pings, props)
+
+"Total of {} pings from client".format(data.count())
+```
+    fetching 223752.06889MB in 118090 files...
+
+
+
+
+
+
+    'Total of 1282154 pings from client'
+
+
+
+
+```python
+counts = data.map(
+    lambda x: (datetime.strptime(x["submission_date"], '%Y%m%d'), 1)
+).countByKey()
+
+pings_counts = [(d,v) for d,v in counts.iteritems() if isinstance(d, datetime)]
```

Comment:
Yes, the counts are of pings, but it's of pings grouped by submission_date. (so long as the date is valid)

How many pings are we ignoring because `not isinstance(d, datetime)`?